src/aws-util/src/s3_uploader.rs

```diff
@@ -0,0 +1,343 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::BTreeMap;
+
+use anyhow::anyhow;
+use aws_sdk_s3::error::SdkError;
+use aws_sdk_s3::operation::complete_multipart_upload::CompleteMultipartUploadError;
+use aws_sdk_s3::operation::create_multipart_upload::CreateMultipartUploadError;
+use aws_sdk_s3::operation::upload_part::UploadPartError;
+use aws_sdk_s3::primitives::ByteStream;
+use aws_sdk_s3::types::{CompletedMultipartUpload, CompletedPart};
+use aws_sdk_s3::Client;
+use aws_types::sdk_config::SdkConfig;
+use bytes::Bytes;
+use bytesize::ByteSize;
+use http::Uri;
+use mz_ore::error::ErrorExt;
+
+/// A multi part uploader which can upload multiple parts
+/// and keeps track of state to eventually finish the upload process.
+/// The caller does not need to know the final number of parts beforehand.
+///
+/// The caller should get an instance by calling `S3MultiPartUploader::new` first.
+/// Then start the multi-part upload by calling `start`.
+/// Each part can be added by calling `upload_part`, and can be called one or more times
+/// and eventually finish the multi part upload by calling `finish` method.
+pub struct S3MultiPartUploader<T = Initialized> {
+    client: Client,
+    bucket: String,
+    path: String,
+    // State to track of ongoing multi part upload.
+    state: T,
+}
+/// The default state just after an `S3MultiPartUploader` is created.
+pub struct Initialized;
+/// State after a multi part upload is in progress.
+pub struct InProgress {
+    upload_id: String,
+    // Mutable state to track the in progress multi part upload.
+    etags: BTreeMap<i32, String>,
+    current_part_number: i32,
+    total_bytes_uploaded: u64,
+}
```

Comment:
It seems a little complex to split out the call to `new` and `start`, since you have to introduce this generic parameter to reflect the two different states. Is there a specific benefit to doing so? ISTM like you might be able to dodge the complexity by just having `async fn new()`, which does what `new` currently does _plus_ what `start` does.

Reply:
ack, will change

Reply:
I had initially started with `S3MultiPartUploader` could be re-used across different paths/uploads. But since this is now tied to a single upload for a s3 path, the states don't make sense anymore. Removed it.